agent/src/main/java/de/metacoder/edwardthreadlocal/AgentEdward.java

```diff
@@ -1,71 +1,30 @@
 package de.metacoder.edwardthreadlocal;
 
-import de.metacoder.edwardthreadlocal.bytecodemanipulation.EdwardTLClassModifier;
+import de.metacoder.edwardthreadlocal.bytecodemanipulation.TestMainClassModification;
+import de.metacoder.edwardthreadlocal.bytecodemanipulation.ThreadLocalClassModification;
+import de.metacoder.edwardthreadlocal.util.IOUtils;
+import de.metacoder.edwardthreadlocal.util.logging.Log;
 
-import java.io.ByteArrayOutputStream;
-import java.io.File;
-import java.io.InputStream;
-import java.lang.instrument.ClassDefinition;
+import java.io.IOException;
 import java.lang.instrument.Instrumentation;
+import java.net.URISyntaxException;
 import java.util.jar.JarFile;
 
-public class AgentEdward {
+import static de.metacoder.edwardthreadlocal.util.logging.Log.info;
 
+public class AgentEdward {
   public static void premain(String agentArgs, Instrumentation inst) throws Exception {
-
-    final EdwardTLClassModifier classModifier = new EdwardTLClassModifier();
-
-    byte[] classBytesFromThreadLocal = loadTL();
-    byte[] patchedBytes = classModifier.patchByteCode(classBytesFromThreadLocal);
-    //byte[] patchedBytes = classBytesFromThreadLocal;
+    Log.DEFAULT_LOGGER.setFineActive(false);
 
     AsciiArt.show();
-
-    // TODO retransform/define here
-
-    System.out.println("Patched byte code size is: " + patchedBytes.length);
-    System.out.println("Loaded " + classBytesFromThreadLocal.length + " bytes from ThreadLocal.class");
-
-    final JarFile agentJar = new JarFile(new File(TraceReceiver.class.getProtectionDomain().getCodeSource().getLocation().toURI()));
-    System.out.println("Attaching agent jar " + agentJar.getName());
-    inst.appendToBootstrapClassLoaderSearch(agentJar);
-
-
-    for(Class<?> clazz : inst.getAllLoadedClasses()) {
-      if("java.lang.ThreadLocal".equals(clazz.getCanonicalName())) {
-        System.out.println("Redefining thread local class!");
-        ClassDefinition classDefinition = new ClassDefinition(clazz, patchedBytes);
-        inst.redefineClasses(classDefinition);
-        System.out.println("Already loaded: " + clazz.getCanonicalName());
-      }
-    }
-
-    inst.addTransformer((loader, className, classBeingRedefined, protectionDomain, classfileBuffer) -> {
-      if("de/metacoder/edwardthreadlocal/test/Main".equals(className)) {
-        System.out.println("Class file transformer now transforming " + className);
-        return new EdwardTLClassModifier().patchByteCode(classfileBuffer);
-      }
-      System.out.println("Class file transformer ignoring saw class " + className);
-      return classfileBuffer;
-    });
-
+    ThreadLocalClassModification.apply(inst);
+    TestMainClassModification.apply(inst);
+    attachAgentJarToBoostrapClassLoaderSearch(inst);
```

Comment:
We should check if other threads may exists when premain is executed. If yes, the instrumentation should be applied after attaching the agent to the bootstrap class loader.